plugins/inputs/t128_graphql/t128_graphql.go

```diff
@@ -0,0 +1,271 @@
+
+	plugin.retrieveMetrics(acc, timestamp)
+
+	return nil
+}
+
+func (plugin *T128GraphQL) retrieveMetrics(acc telegraf.Accumulator, timestamp time.Time) {
+	request, err := plugin.createRequest()
+	if err != nil {
+		acc.AddError(fmt.Errorf("failed to create a request for query %s: %w", plugin.Query, err))
+		return
+	}
+
+	response, err := plugin.client.Do(request)
+	if err != nil {
+		acc.AddError(fmt.Errorf("failed to make graphQL request for collector %s: %w", plugin.CollectorName, err))
+		return
+	}
+	defer response.Body.Close()
+
+	message, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		message = []byte("")
+	}
+
+	if response.StatusCode < 200 || response.StatusCode >= 300 {
+		template := fmt.Sprintf("status code %d not OK for collector ", response.StatusCode) + plugin.CollectorName + ": %s"
+		for _, err := range decodeAndReportJSONErrors(message, template) {
+			acc.AddError(err)
+		}
+		return
+	}
+
+	//decode json
+	jsonParsed, err := gabs.ParseJSON(message)
+	if err != nil {
+		acc.AddError(fmt.Errorf("invalid json response for collector %s: %w", plugin.CollectorName, err))
+		return
+	}
+
+	jsonObj, err := jsonParsed.JSONPointer(plugin.JSONEntryPoint)
+	if err != nil {
+		template := "unexpected response for collector " + plugin.CollectorName + ": %s"
+		for _, err := range decodeAndReportJSONErrors(message, template) {
+			acc.AddError(err)
+		}
+		return
+	}
+
+	//update acc
+	jsonChildren, err := jsonObj.Children()
+	if err != nil {
+		acc.AddError(fmt.Errorf("failed to iterate on response nodes for collector %s: %w", plugin.CollectorName, err))
+		return
+	}
+
+	plugin.processResponse(jsonChildren, acc, timestamp)
+}
+
+func (plugin *T128GraphQL) processResponse(jsonChildren []*gabs.Container, acc telegraf.Accumulator, timestamp time.Time) {
+	//TODO: allow nested fields/tags - MON-310
+	for _, child := range jsonChildren {
+		node := child.Data().(map[string]interface{})
+		fields := make(map[string]interface{})
+		tags := make(map[string]string)
+
+		for fieldRenamed, fieldName := range plugin.Fields {
+			if isNil(node[fieldName]) {
+				acc.AddError(fmt.Errorf("found empty data for collector %s: field %s", plugin.CollectorName, fieldName))
+				continue
+			}
+			fields[fieldRenamed] = node[fieldName]
+		}
+
+		for tagRenamed, tagName := range plugin.Tags {
+			if isNil(node[tagName]) {
+				tags[tagRenamed] = ""
+				continue
+			}
+			tags[tagRenamed] = fmt.Sprintf("%v", node[tagName])
+		}
+
+		acc.AddFields(
+			plugin.CollectorName,
+			fields,
+			tags,
+			timestamp,
+		)
+	}
+}
+
+func (plugin *T128GraphQL) createRequest() (*http.Request, error) {
+	content := struct {
+		Query string `json:"query,omitempty"`
+	}{
+		plugin.Query,
+	}
+
+	//inject env config?
+	body, err := json.Marshal(content)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request body for query '%s': %w", plugin.Query, err)
+	}
+
+	request, err := http.NewRequest("POST", plugin.BaseURL, bytes.NewReader(body))
```

Comment:
Everything up to this point (I guess the creation of the request body) could be done in `Init` rather than being repeated for every request. I'm not sure whether you could potentially create the full `http.Request` and just reuse the fully populated object. I expect there would be issues with that.